File: sdks/cpp/examples/flatfiles_trade_quotes_to_arrow.cpp
// FLATFILES dynamic-schema decode -> Arrow IPC bytes.
//
// The flat-file surface returns whole-universe data for a single
// (sec_type, req_type, date) tuple. Decoded shape is determined at
// runtime by the request type, so the C++ wrapper exposes Arrow IPC
// bytes — pair with arrow-cpp on the consumer side to materialise an
// `arrow::Table`.

#include <fstream>
#include <iostream>

#include "thetadx.hpp"

int main() {
    try {
        auto creds = thetadatadx::Credentials::from_file("creds.txt");
        auto config = thetadatadx::Config::production();
        auto unified = thetadatadx::Client::connect(creds, config);

        // Whole-universe option trade-quotes for one trading day.
        auto rows = unified.flat_files().option_trade_quote("20260428");
        std::cout << "option_trade_quote rows: " << rows.size() << std::endl;

        // Arrow IPC stream bytes -- feed into arrow::ipc::RecordBatchStreamReader
        // to materialise a typed RecordBatch with the dynamic schema.
        auto ipc = rows.to_arrow_ipc();
        std::ofstream out("/tmp/option-trade-quote.arrow", std::ios::binary);
        out.write(reinterpret_cast<const char*>(ipc.data()),
                  static_cast<std::streamsize>(ipc.size()));
        std::cout << "wrote " << ipc.size() << " bytes of Arrow IPC to "
                  << "/tmp/option-trade-quote.arrow" << std::endl;

        // Same path, dispatched dynamically.
        auto oi = unified.flat_files().request("OPTION", "OPEN_INTEREST", "20260428");
        std::cout << "open_interest rows: " << oi.size() << std::endl;

        // Drop raw vendor CSV bytes to disk without materialising rows.
        unified.flat_files().to_path("OPTION", "TRADE_QUOTE", "20260428",
                                     "/tmp/option-trade-quote", "csv");
        std::cout << "raw vendor CSV at /tmp/option-trade-quote.csv" << std::endl;
    } catch (const std::exception& e) {
        std::cerr << "error: " << e.what() << std::endl;
        return 1;
    }
    return 0;
}

File: sdks/typescript/examples/flatfiles_trade_quotes_to_arrow.ts
// FLATFILES dynamic-schema decode -> Arrow IPC -> apache-arrow Table.
//
// The flat-file surface returns whole-universe data for a single
// (SecType, ReqType, date) tuple. Decoded shape is determined at
// runtime by the request type, so the binding emits Arrow IPC bytes
// and the caller deserialises with `apache-arrow`'s `tableFromIPC`.
//
// Run with: `npx tsx flatfiles_trade_quotes_to_arrow.ts`
// Requires: `npm i apache-arrow` (peer dep, not bundled by thetadatadx).

import { Client } from "thetadatadx";
import { tableFromIPC } from "apache-arrow";

async function main(): Promise<void> {
  const client = await Client.connectFromFile("creds.txt");

  // Whole-universe option trade-quotes for one trading day.
  const rows = client.flatFiles.optionTradeQuote("20260428");
  console.log(`option_trade_quote rows: ${rows.len()}`);

  // Apache Arrow table -- one column per vendor field plus the contract
  // key columns (symbol, expiration, strike, right). Schema inferred
  // from the first row by `flatfiles::arrow::rows_to_arrow`.
  const ipc = rows.toArrowIpc();
  const table = tableFromIPC(ipc);
  console.log(table.schema.fields.map((f) => `${f.name}:${f.type}`).join(", "));

  // Same path, dispatched dynamically.
  const oi = client.flatFiles.request("OPTION", "OPEN_INTEREST", "20260428");
  console.log(`open_interest rows: ${oi.len()}`);

  // Drop raw vendor CSV bytes to disk without materialising rows.
  const path = client.flatFileToPath(
    "OPTION",
    "TRADE_QUOTE",
    "20260428",
    "/tmp/option-trade-quote",
    "csv",
  );
  console.log(`raw vendor CSV at ${path}`);
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
